src/java/htsjdk/samtools/sra/SRALazyRecord.java

```diff
@@ -0,0 +1,941 @@
+    }
+
+    private boolean getReadPairedFlagImpl() {
+        try {
+            if (isAligned) {
+                return getCurrentAlignment().isPaired();
+            } else {
+                return getCurrentUnalignedRead().getNumFragments() > 1;
+            }
+        } catch (ErrorMsg e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private boolean getNotPrimaryAlignmentFlagImpl() {
+        try {
+            if (isAligned) {
+                return getCurrentAlignment().getAlignmentCategory() == Alignment.secondaryAlignment;
+            }
+        } catch (ErrorMsg e) {
+            throw new RuntimeException(e);
+        }
+
+        return false;
+    }
+
+    private boolean getMateNegativeStrandFlagImpl() {
+        try {
+            if (isAligned && getReadPairedFlag() && !getMateUnmappedFlag()) {
+                Alignment mate = getCurrentAlignment().getMateAlignment();
+                return mate.getIsReversedOrientation();
+            }
+        } catch (ErrorMsg e) {
+            throw new RuntimeException(e);
+        }
+
+        return false;
+    }
+
+    private boolean getMateUnmappedFlagImpl() {
+        try {
+            if (isAligned) {
+                return !getCurrentAlignment().hasMate();
+            } else {
+                Read unalignedRead = getCurrentUnalignedRead();
+                int numFragments = unalignedRead.getNumFragments();
+                int nextFragmentIdx = unalignedReadFragmentIndex + 1;
+                if (nextFragmentIdx == numFragments) {
+                    nextFragmentIdx = 0;
+                }
+
+                return unalignedRead.fragmentIsAligned(nextFragmentIdx);
+            }
+        } catch (ErrorMsg e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private boolean getFirstOfPairFlagImpl() {
+        if (!getReadPairedFlag()) {
+            return false;
+        }
+        try {
+            if (isAligned) {
+                String fragmentId = getCurrentAlignment().getFragmentId();
+                if (!fragmentId.contains(".FA")) {
+                    throw new RuntimeException("Invalid fragment id: " + fragmentId);
+                }
+
+                return fragmentId.contains(".FA0.");
+            } else {
+                return unalignedReadFragmentIndex == 0;
+            }
+            // TODO: check if we can forget about more than two records
+        } catch (ErrorMsg e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private boolean getSecondOfPairFlagImpl() {
+        if (!getReadPairedFlag()) {
+            return false;
+        }
+        try {
+            if (isAligned) {
+                String fragmentId = getCurrentAlignment().getFragmentId();
+                if (!fragmentId.contains(".FA")) {
+                    throw new RuntimeException("Invalid fragment id: " + fragmentId);
+                }
+
+                return !fragmentId.contains(".FA0.");
```

Comment:
Declare ".FA" as a named constant at the top of the class (`FRAGMENT_ID_PREFIX`?)